### PR TITLE
Fix temp directory permission issue on worker side

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

(Details)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/file_utils.py
Comment (BenWilson2) [inline mlflow/utils/file_utils.py:506]: do we need the workers to have write authority over this location? Could we do 0o755 instead?

Comment (BenWilson2) [reply mlflow/utils/file_utils.py:506]: nevermind. I just checked the docs on `os.makedirs()` and the *default* behavior there is 0o777. 
tempfile.mkdtemp sets 0o700 by default. 
This looks good to me!

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Comment by WeichenXu123** [comment]
Let me add some comments before merging

File: mlflow/utils/file_utils.py
Comment (harupy) [inline mlflow/utils/file_utils.py:506]:
```suggestion
        # mkdtemp creates a directory with permission 0o700
        # https://github.com/python/cpython/blob/0859368335d470b9ff33fc53ed9a85ec2654b278/Lib/tempfile.py#L368
        # Change it to 0o777 to ensure it can be seen in spark UDF
```

File: mlflow/utils/file_utils.py
Comment (harupy) [inline mlflow/utils/file_utils.py:533]: Does the temporary directory have to be writable?

Comment (WeichenXu123) [reply mlflow/utils/file_utils.py:533]: Because "shutils" makedirs also use "0o777", I keep it the same with it

Comment (harupy) [reply mlflow/utils/file_utils.py:533]: So it has to be writable?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !

File: mlflow/utils/file_utils.py
Comment (dbczumar) [reply mlflow/utils/file_utils.py:533]: I don't think it strictly needs to be writeable, but this was the preexisting behavior for the Spark model cache. To avoid regressions, I think it's best to use the same permissions level.

Comment (harupy) [reply mlflow/utils/file_utils.py:533]: Makes sense! I asked this because `can be seen` sounds like `should be readable but doesn't have to be writable`.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@WeichenXu123 Before merging, can we resolve `os.makedirs()` permissions exceptions thrown when operating on Table ACL clusters?

File: mlflow/utils/nfs_on_spark.py
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:34]: Do we remove this line before the merge?

Comment (harupy) [reply mlflow/utils/nfs_on_spark.py:34]: Ah ok

Comment (dbczumar) [reply mlflow/utils/nfs_on_spark.py:34]: We'll keep this line because we're now testing whether or not we can write to NFS by attempting to create a directory at `test_path`

Comment (harupy) [reply mlflow/utils/nfs_on_spark.py:34]: Can we use `os.access(nfs_root_dir, os.W_OK)`?

ref: https://stackoverflow.com/a/2113511/6943581

Comment (dbczumar) [reply mlflow/utils/nfs_on_spark.py:34]: I think we should try to write, as is done in the current PR. The best way to test for write access is to try to write.

From the comments of https://stackoverflow.com/a/2113511/6943581:

```
Testing a directory for just the write bit isn't enough if you want to write files to the directory. You will need to test for the execute bit as well if you want to write into the directory. os.access('/path/to/folder', os.W_OK | os.X_OK) With os.W_OK by itself you can only delete the directory (and only if that directory is empty
```

If we're not careful, we might make a mistake.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks a bunch, @WeichenXu123 !

File: mlflow/utils/nfs_on_spark.py
Comment (harupy) [reply mlflow/utils/nfs_on_spark.py:34]: Sounds good!

File: mlflow/utils/nfs_on_spark.py
Comment (harupy) [inline mlflow/utils/nfs_on_spark.py:43]: Does this line fail when `test_path` doesn't exist?

Comment (dbczumar) [reply mlflow/utils/nfs_on_spark.py:43]: Good catch! Fortunately, it swallows:

```
In [11]: shutil.rmtree("/tmp/asfsadfsdf", ignore_errors=True)

In [12]: shutil.rmtree("/tmp/asfsadfsdf")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-12-8bc109b15d91> in <module>
----> 1 shutil.rmtree("/tmp/asfsadfsdf")

~/opt/anaconda3/lib/python3.8/shutil.py in rmtree(path, ignore_errors, onerror)
    704             orig_st = os.lstat(path)
    705         except Exception:
--> 706             onerror(os.lstat, path, sys.exc_info())
    707             return
    708         try:

~/opt/anaconda3/lib/python3.8/shutil.py in rmtree(path, ignore_errors, onerror)
    702         # lstat()/open()/fstat() trick.
    703         try:
--> 704             orig_st = os.lstat(path)
    705         except Exception:
    706             onerror(os.lstat, path, sys.exc_info())

FileNotFoundError: [Errno 2] No such file or directory: '/tmp/asfsadfsdf'
```

Comment (harupy) [reply mlflow/utils/nfs_on_spark.py:43]: Confirmed it doesn't

**Review by harupy — APPROVED** [review]

**Pull request merged**
